src/main/java/app/linguistai/bmvp/service/gamification/quest/QuestService.java

```diff
@@ -56,8 +77,162 @@ public Boolean checkUserHasQuestType(String email, QuestCompletionCriteria type)
+        }
+
+        return useWordQuests;
+    }
+
+    /**
+     * Select a random word from the retrieved unknown word list, while ensuring no duplicates exist for quests
+     * @param listId List ID for which the words will be pulled from
+     * @param usedWordsSoFar Used words so far in the generation
+     * @return Random unique word for "Use Word" type quests
+     * @throws Exception When a random word cannot be retrieved from list with ID listId
+     */
+    private String returnUniqueRandomWordForUseWordQuest(UUID listId, HashSet<String> usedWordsSoFar) throws Exception {
+        final int HARDCODED_DO_WHILE_LIMIT = 100; // to prevent infinite loops
+        int currentIteration = 0;
+        String randomWord;
+
+        do {
+            randomWord = unknownWordService.getRandomWordFromList(listId);
+            currentIteration++;
+        } while (usedWordsSoFar.contains(randomWord) && currentIteration < HARDCODED_DO_WHILE_LIMIT);
+
+        return randomWord;
+    }
+
+    private Quest buildUseWordQuest(User assignee, String word, Integer times) {
+        return Quest.builder()
+            .user(assignee)
+            .title(WORD_PRACTICE_TITLE)
+            .description(this.generateUseWordQuestDesc(word, times))
+            .type(TYPE_USE_WORD)
+            .image(QUEST_WORD_IMAGE)
+            .reward(BASE_USE_WORD_REWARD * times)
+            .assignedDate(DateUtils.convertUtilDateToSqlDate(Calendar.getInstance().getTime()))
+            .completionCriteria(new UseWordCriteria(word, times))
+            .build();
+    }
+
+    private Quest buildSendMessageQuest(User assignee, Integer times) {
+        return Quest.builder()
+            .user(assignee)
+            .title(SEND_MESSAGE_TITLE)
+            .description(this.generateSendMessageQuestDesc(times))
+            .type(TYPE_SEND_MESSAGE)
+            .image(QUEST_MESSAGE_IMAGE)
+            .reward(BASE_SEND_MESSAGE_REWARD * times)
+            .assignedDate(DateUtils.convertUtilDateToSqlDate(Calendar.getInstance().getTime()))
+            .completionCriteria(new SendMessageCriteria(times))
+            .build();
+    }
+
+    private Boolean userHasActiveQuests(User user) {
+        List<Quest> userQuests = questRepository.findAllByUserId(user.getId());
+        java.sql.Date today = DateUtils.convertUtilDateToSqlDate(Calendar.getInstance().getTime());
+        return userQuests.stream().anyMatch(quest -> Objects.equals(quest.getAssignedDate(), today));
+    }
+
+    private void deleteInactiveQuests(User user) {
+        // Get all user quests
+        List<Quest> userQuests = questRepository.findAllByUserId(user.getId());
+
+        // Get today's date in java.sql.Date format
+        java.sql.Date today = DateUtils.convertUtilDateToSqlDate(Calendar.getInstance().getTime());
+
+        // Remove all quests assigned today
+        userQuests.removeIf(quest -> quest.getAssignedDate().equals(today));
```

Comment:
I didn't understand this part - are we not supposed to remove the quests that were assigned before today (i.e., inactive)

Reply:
Maybe I can add a comment to clarify. We remove the active quests from the `List<Quest> userQuests`, so that when we `deleteAll(userQuests)` later, active quests are **not deleted**.